### PR TITLE
RD-3016 Add hiding `Generate in the Composer` button feature.

### PR DESCRIPTION
**Pull request by stepek**
## Description
Added support for hiding `Generate in the Composer` button

## Screenshots / Videos
image

## Checklist

- [X] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [X] I verified if that change requires a documentation update.
- [X] I added proper labels to this PR.

## Tests
[#1149 (27-Aug-2021 13:06:16)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1149/)
[Build #1154 (30-Aug-2021 07:57:59)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1154/)
[ Build #1164 (31-Aug-2021 08:32:56)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1164/)
[Build #1173 (01-Sep-2021 09:27:36)](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/1173/) PASS

 

## Documentation
Will be done in https://cloudifysource.atlassian.net/browse/RD-3069

### CODE REVIEW

File: app/translations/en.json [hunk 1]
```diff
@@ -690,7 +690,7 @@
                     }
                 },
                 "hideFailedBlueprints": "Hide failed blueprints",
-                "showEditCopyInComposerButton": "Show the \"Edit a copy in Composer\" button",
+                "showComposerOptions": "Show the Composer options",
```
[the remark this answers is not in the document]
Comment (qooban) [reply app/translations/en.json:693]: I'm unresolving this conversation as it seems like not addressed, right?

File: widgets/common/src/BlueprintUploadActionsMenu.tsx [hunk 2]
```diff
@@ -8,18 +8,27 @@ const menuItems: ComponentProps<typeof Stage.Basic.Menu.Item>[] = [
     { name: 'generateInComposer', key: 'generateInComposer', content: t('generateInComposer') }
 ];
 
+const getMenuItems = (includeComposerButton: boolean) => {
+    if (includeComposerButton) {
+        return menuItems;
+    }
+    return menuItems.filter(item => item.name !== 'generateInComposer');
```
[the remark this answers is not in the document]
Comment (stepek) [reply widgets/common/src/BlueprintUploadActionsMenu.tsx:15]: ok

**Review by qooban — COMMENTED** [review]
I don't have specific comments other than those left by Kuba M.

Please check system tests results (seems like Blueprints widget spec is failing) and CI (TS error in `test/cypress/integration/widgets/blueprints_spec.ts`).

File: app/translations/en.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (stepek) [reply app/translations/en.json]: I resolving before push

**Comment by qooban** [comment]
@stepek Please check CI.

File: widgets/common/src/BlueprintUploadActionsMenu.tsx [hunk 3]
```diff
@@ -4,15 +4,17 @@ import type { MarketplaceTab } from './blueprintMarketplace/types';
...
+    { name: 'uploadFromPackage', key: 'uploadFromPackage', content: t('uploadFromPackage') }
 ];
 
 const getMenuItems = (includeComposerButton: boolean) => {
     if (includeComposerButton) {
-        return menuItems;
+        return [
+            ...menuItems,
+            { name: 'generateInComposer', key: 'generateInComposer', content: t('generateInComposer') }
+        ];
```
Comment (kubama) [inline widgets/common/src/BlueprintUploadActionsMenu.tsx]: For consistency and better readability the menu item added in L14 should be also a top level const placed next to the `menuItems` declaration. I would also rename `menuItems` to something reflecting that these are not all possible menu options - `basicMenuItems`, `baseMenuItems` or something similar.

**Review by kubama — DISMISSED** [review]

**Review by qooban — APPROVED** [review]

File: test/cypress/integration/widgets/blueprints_spec.ts [hunk 4]
```diff
@@ -59,7 +59,7 @@ describe('Blueprints widget', () => {
 
         it('should not show the "Edit a copy in Composer" button if it is turned off in the configuration', () => {
             cy.editWidgetConfiguration('blueprints', () => {
-                cy.contains('.field', 'Show the "Edit a copy in Composer" button')
+                cy.contains('.field', 'Show Composer options')
```
Comment (qooban) [inline test/cypress/integration/widgets/blueprints_spec.ts:62]: Hmm... The change is fine, but I've just realized: do we have a test for `Generate in the Composer` button?

If not, I think it is worth to add to check if:
1. It opens Composer (just URL check)
2. It respects "Show Composer options" configuration field

**Review by qooban — APPROVED** [review]
Added one comment.

File: test/cypress/integration/widgets/blueprints_spec.ts (hunk 4 above)
Comment (stepek) [reply test/cypress/integration/widgets/blueprints_spec.ts:62]: Will add

File: test/cypress/integration/widgets/blueprints_spec.ts [hunk 5]
```diff
@@ -420,7 +419,36 @@ describe('Blueprints widget', () => {
...
+            cy.editWidgetConfiguration('blueprints', () => {
+                cy.contains('.field', 'Show Composer options')
+                    .find('input')
+                    // NOTE: force, as the checkbox from Semantic UI is
+                    // class=hidden which prevents Cypress from clicking it
+                    .click({ force: true });
+            });
+
+            cy.contains('Upload').click();
+            cy.contains('Generate in the Composer').should('not.exist');
```
Comment (kubama) [inline test/cypress/integration/widgets/blueprints_spec.ts:449]: Maybe worth to assert that other options are visible?

File: test/cypress/integration/widgets/blueprints_spec.ts [hunk 6]
```diff
@@ -420,7 +419,36 @@ describe('Blueprints widget', () => {
         });
     });
 
+    describe('should open Composer', () => {
+        before(() => {
+            cy.reload();
```
Comment (kubama) [inline test/cypress/integration/widgets/blueprints_spec.ts:424]: Why full reload? Isn't `cy.refreshPage` sufficient?

File: test/cypress/integration/widgets/blueprints_spec.ts [hunk 7]
```diff
@@ -420,7 +419,36 @@ describe('Blueprints widget', () => {
...
+            cy.contains('Upload').click();
+            cy.contains('Generate in the Composer').should('be.visible');
+
+            cy.editWidgetConfiguration('blueprints', () => {
+                cy.contains('.field', 'Show Composer options')
+                    .find('input')
+                    // NOTE: force, as the checkbox from Semantic UI is
+                    // class=hidden which prevents Cypress from clicking it
+                    .click({ force: true });
+            });
```
Comment (qooban) [inline test/cypress/integration/widgets/blueprints_spec.ts:446]: I think you can use custom command I recently added and have here a bit simpler version:
```suggestion
            cy.setBooleanConfigurationField('blueprints', 'Show Composer options', true);
```
Will that work?

File: test/cypress/integration/widgets/blueprints_spec.ts [hunk 8]
```diff
@@ -420,7 +419,36 @@ describe('Blueprints widget', () => {
         });
     });
 
+    describe('should open Composer', () => {
+        before(() => {
+            cy.reload();
+        });
+
+        it('after click menu item', () => {
```
Comment (qooban) [inline test/cypress/integration/widgets/blueprints_spec.ts:427]: To keep test naming consistent (with e.g. "should open Composer with the blueprint imported on "Edit a copy in Composer" icon click"), I'd suggest to change it to:
```suggestion
        it('on "Generate in the Composer" menu item click', () => {
```

**Review by qooban — COMMENTED** [review]
Thanks for adding new tests. Added some comments. Should be quick to address.

File: test/cypress/integration/widgets/blueprints_spec.ts (hunk 6 above)
Comment (stepek) [reply test/cypress/integration/widgets/blueprints_spec.ts:424]: using `refreshPage()` `cy.window().its(open)` don't work correctly, this was way it work properly again what i found.

File: test/cypress/integration/widgets/blueprints_spec.ts (hunk 5 above)
Comment (stepek) [reply test/cypress/integration/widgets/blueprints_spec.ts:449]: IMHO better in this place with other config-change test cases

**Review by qooban — APPROVED** [review]

File: test/cypress/integration/widgets/blueprints_spec.ts (hunk 5 above)
Comment (kubama) [reply test/cypress/integration/widgets/blueprints_spec.ts:443]: What I suggested is adding checks for the other two options that should be present in the menu after clicking the 'Upload' button. This way we would check that the actual menu is displayed, it's just the 'Generate in the Composer' option that is missing. With the current implementation we are only checking if 'Generate in the Composer' is absent, which will be true even if the actual dropdown menu is not displayed.

Comment (kubama) [reply test/cypress/integration/widgets/blueprints_spec.ts:443]: Looks like my comment was not addressed anyhow 🙁.

Comment (stepek) [reply test/cypress/integration/widgets/blueprints_spec.ts:443]: i miss this comment before merge, sorry about that, will add check for that

Comment (stepek) [reply test/cypress/integration/widgets/blueprints_spec.ts:443]: Added in https://github.com/cloudify-cosmo/cloudify-stage/pull/1583

Comment (kubama) [reply test/cypress/integration/widgets/blueprints_spec.ts:443]: Thanks! 👍

**Pull request merged**